### ISSUE TITLE
Remote Drive Library

### ISSUE THREAD

**Issue by superdan-t** [BinghamtonRover/BurtOS-2#35]
## Objective
Build a remote control library for the base station and autonomy programs to drive the rover remotely by sending network messages. This includes combining inputs from multiple sources: the base station will produce two desired speeds: one from keyboard bindings, one from controller bindings.
## Strategy & Requirements
* Encapsulate these requirements in one class (plus potential support classes, all in one header)
  * Make class in a new folder called (rover_control, remote, remote_rover, rover_remote_control, remote_control, rc, rctrl?)
    * Names of the class, namespace, etc. are flexible. I'm not sure I like any of the ones I came up with
  * Will be a new CMake target too
* Send movement packets at a regular interval. Use a default of 100 milliseconds, but create setters/getters for the interval. Drive library users will have to call an update function in the main loop for checking this (something like `Drive::poll_events()`)
* ~~Each input source produces a pair: speed (float) and steering angle (float) (Could use a support class/struct with only those two fields for this). The Drive module will have to save a collection of pointers to these input sources, but must select only one to send. Selection criteria: Select the highest magnitude speed and highest magnitude steering independently~~
  * ~~For saving input sources, expose API functions like `add_input_source(const Drive::MovementInput& src)`. In the implementation, you'll probably need a vector of pointers (they can be pointers to const objects as the drive module should only be reading the input sources).~~
  * ~~This may need more discussion (also use your judgement). Saving pointers is probably not good since this class isn't managing their lifetime. Maybe use a vector of MovementInputs, and the add function returns an index. Usage examples shown below~~
* Provide API functions for the other drive messages (set drive mode, halt, etc.)
  * Update: set speed/steering/movement (combination of both) should have identical API functions except that movement messages are not sent with every call. Movement messages are sent at an interval. The other message types (which are more like "commands" than movement updates), should immediately send the message.
* Require that the drive remote controller is constructed with a reference to a network sender `net::MessageSender&`. It is expected that this sender is configured to send to the base station upon construction.
## ~~Example of Desired Usage~~
```C++
// Assume to_subsystem is a net::MessageSender configured to send to the subsystem program

rc::Drive remote_drive(to_subsystem);

/*
 Example with pointers
*/

// Keyboard bindings will update the speed in this object
rc::Drive::MovementInput keyboard;
// Controller bindings will update this one
rc::Drive::MovementInput controller;

// Now, remote_drive saves a pointer to keyboard/controller, but it doesn't manage their lifetime
// They could go out of scope and we have a dangling pointer
// Easy for users to make mistakes
remote_drive.add_input_source(keyboard);
remote_drive.add_input_source(controller);

// Setting speed example
keyboard.set_speed(1.0F);

/*
  Example with index
*/

std::size_t keyboard_input_source = remote_drive.add_input_source();
std::size_t controller_input_source = remote_drive.add_input_source();

// Setting speed example. Not as pretty as above
remote_drive.input_sources()[keyboard_input_source].set_speed(1.0F);

for (;;) {
    // program stuff ...

    remote_drive.poll_events();

    // more program stuff ...
}
```
## Tips
* The name scheme isn't super important right now since it can be changed easily
* Use normal OOP standards for encapsulation
  * private variables, controlled access through getters/setters, you can do inline getters/setters (defined in header) for concision (rover team has no defined standard for this, but I personally feel that separating implementation/specification for a 1 line function isn't worth it)

**Comment by superdan-t** [comment]
I'd like to update this to remove the multiple input sources requirement. The remote drive object itself should have set speed/angle functions.

I think that trying to solve the "keyboard or controller" problem by supporting multiple input sources in the drive library is a bit of over-engineering. The base station should select between them before calling `RemoteDrive::set_speed` and such.

**Comment by superdan-t** [comment]
Closed by #37
